### PR TITLE
[LLT-5972] Control Up/Down state of WG-NT adapter according to presence of peers.

### PR DESCRIPTION
**Pull request by Jauler**
### Problem
It may be beneficial to show the adapter as "disconnected" or "down" in case it is idling in libtelio. This is especially useful if one keeps it preconfigured for potential use in the future, as adapter configuration can sometimes take seconds. For this reason this PR adds a logic which makes sure that adapter is down if there are no peers requested for WG and brings it up when peers are added. 

There is RFC LLT-0089 for more details.

Now in order to finalize the PR there had to be a few changes to other places.

### Windows IP address setup
Previously we have been configuring windows adapter IP address with standard `netsh` command and later waiting until the IP address is actually set, Unfortunately due to a fact that this operation now is operating on "disconnected" adapter before any feature is turned on - the waiting would be indefinite, as until any feature is enabled, the adapter will not get its IP. This waiting was added due to a fact that we have observed that it may take seconds for the IP address to be actually set. The delay was coming from windows duplicate address detection, but duplicate address detection on tunnel interface is not useful. For this reason in this PR, the DAD was disabled and waiting for IPs removed.

### Consolidating WG port for proxy and starcast virtual peer
Previously assumption was made that WireGuard listen port never changes during runtime. While this assumption is mostly true, it is a bit dangerous, and out right broken in case we are starting with disabled adapter. For this reason in this PR a runtime consolidation step to align listen port used by `telio-starcast` and `telio-proxy` is added on wg_controller consolidation cycle.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

**Comment by ThrasherLT** [comment]
+1.0

File: crates/telio-wg/src/adapter/windows_native_wg.rs [hunk 1]
```diff
@@ -187,6 +175,60 @@ impl WindowsNativeWg {
...
+    /// `ensure_adapter_state` attempts to bring interface up (requested_state == true) if
+    /// it is requested to be up and is not already up. Or bring it down if it is requested
+    /// to be down, but adapter is not yet down.
+    ///
+    /// As one may expect - this function is idempotent.
+    async fn ensure_adapter_state(&self, requested_state: bool) -> Result<(), AdapterError> {
+        // We have seen a few cases where single attempt to bring
+        // The adapter up caused some issues. Especially if performed
+        // Early after driver installation.
+        // Exact root cause is not yet found, therefore lets attempt to bring
```
Comment (lcruz99) [inline crates/telio-wg/src/adapter/windows_native_wg.rs:189]: Is there any ticket for this investigation/issue? If so, please add the ticket # here

File: crates/telio-wg/src/adapter/windows_native_wg.rs [hunk 2]
```diff
@@ -187,6 +175,60 @@ impl WindowsNativeWg {
...
+        // The interface up for slightly longer before declaring error.
+        let state_str = if requested_state { "up" } else { "down" };
+        let mut os_error = IOError::from_raw_os_error(0);
+        for _ in 0..5 {
+            // Retreive up/down state of the adapter
```
Comment (lcruz99) [inline crates/telio-wg/src/adapter/windows_native_wg.rs:194]: nit: Retreive -> Retrieve

File: src/device/wg_controller.rs [hunk 3]
```diff
@@ -183,6 +190,27 @@ async fn consolidate_wg_private_key<W: WireGuard>(
     Ok(())
 }
 
+async fn consolidate_wg_listen_port<W: WireGuard, P: Proxy>(
+    _: &RequestedState,
```
Comment (mathiaspeters) [inline src/device/wg_controller.rs:194]: Is there are reason this is here when it's unused?

**Comment by lcruz99** [comment]
Should we test the interface state/config when there's no peers?

File: src/device/wg_controller.rs [hunk 4]
```diff
@@ -183,6 +190,27 @@ async fn consolidate_wg_private_key<W: WireGuard>(
...
+    wireguard_interface: &W,
+    proxy: Option<&P>,
+    starcast_vpeer: Option<&Arc<StarcastPeer>>,
+) -> Result {
+    if let Some(proxy) = proxy {
+        let listen_port = wireguard_interface.get_interface().await?.listen_port;
+        let wg_addr = listen_port.map(|p| (Ipv4Addr::LOCALHOST, p).into());
+        proxy.set_wg_address(wg_addr).await?;
```
Comment (LukasPukenis) [inline src/device/wg_controller.rs:202]: Question - if the adapter is `Down` does it report any `Listen Port` via UAPI? I assume it's not there but want to be sure :thinking:

File: src/device/wg_controller.rs [hunk 5]
```diff
@@ -183,6 +190,27 @@ async fn consolidate_wg_private_key<W: WireGuard>(
...
+    }
+
+    if let Some(starcast_vpeer) = starcast_vpeer {
+        let listen_port = wireguard_interface.get_interface().await?.listen_port;
+        let wg_addr = listen_port.map(|p| (Ipv4Addr::LOCALHOST, p).into());
+        starcast_vpeer.set_wg_address(wg_addr).await?;
+    }
```
Comment (LukasPukenis) [inline src/device/wg_controller.rs:209]: This also would be great to have unit-tested and that shouldn't be too difficult given that all relevant parts are already interfaces. If possible, I would like to have this unit tested or at least natlab-tested (whichever is easier)

File: crates/telio-wg/src/adapter/windows_native_wg.rs [hunk 6]
```diff
@@ -208,22 +250,41 @@ impl Adapter for WindowsNativeWg {
...
+                            interface: None,
+                        }),
+                        None,
+                    ),
+                };
 
-                    Ok(self.get_config_uapi())
+                // If all of the peers has been removed -> bring the adapter down
+                if peer_cnt.map(|p| p == 0).unwrap_or(false) {
+                    self.ensure_adapter_state(false).await?;
```
Comment (LukasPukenis) [inline crates/telio-wg/src/adapter/windows_native_wg.rs:283]: Opinion: it makes sense that if the source of the truth - peer_cnt is present we use that to set the state and if not we're left with `set_cfg.peers` however I think it would be clearer to do that once, at the end of the function which would clearly show the state machine and reasoning as it might be easy to miss one of the two `ensure_adapter_state()`.

Nitpick: Could we add enum(Down, Up) instead of bools :bow:

File: src/device/wg_controller.rs (hunk 4 above)
Comment (Jauler) [reply src/device/wg_controller.rs:202]: The WG-NT behavior is rather strange, it reports a listen port, but it changes after going `Up`.

This is why I decided to add WG listen port consolidation in `wg_controller` such that we would not just assume that WG listen port never changes.

File: src/device/wg_controller.rs (hunk 5 above)
Comment (Jauler) [reply src/device/wg_controller.rs:209]: This was in-fact natlab finding :stuck_out_tongue: So it is being tested. But I missed direct testcase for the interface control itself, although I remember writing it.

I will add that one.

File: crates/telio-wg/src/adapter/windows_native_wg.rs (hunk 6 above)
Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs:283]: >  however I think it would be clearer to do that once, at the end of the function which would clearly show the state machine and reasoning as it might be easy to miss one of the two ensure_adapter_state()

One must set the interface up before configuring peers. And in a mirroring fashion - it is considered "clean" to remove all of them before bringing interface down (in other words, any peer lifetime should be fully covered with interface being up). This is why `Down->Up` transition is handled before executing `Set`, and `Up->Down` after executing `Set` command

Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs:283]: > Nitpick: Could we add enum(Down, Up) instead of bools 🙇

That is a good one, Will do.

File: crates/telio-wg/src/adapter/windows_native_wg.rs (hunk 2 above)
Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs]: Fixed

File: crates/telio-wg/src/adapter/windows_native_wg.rs (hunk 1 above)
Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs:195]: Fixed

File: crates/telio-wg/src/adapter/windows_native_wg.rs [hunk 7]
```diff
@@ -187,6 +175,60 @@ impl WindowsNativeWg {
...
     }
 
+    /// Ensure adapter state (up/down)
+    ///
+    /// `ensure_adapter_state` attempts to bring interface up (requested_state == true) if
+    /// it is requested to be up and is not already up. Or bring it down if it is requested
+    /// to be down, but adapter is not yet down.
+    ///
+    /// As one may expect - this function is idempotent.
+    async fn ensure_adapter_state(&self, requested_state: bool) -> Result<(), AdapterError> {
```
[the remark this answers is not in the document]
Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs]: I have introduced `AdapterState` enum, which have members `Up` and `Down` instead of a `bool`. Also changed `requested/actual` to `want/have`.

**Comment by Jauler** [comment]
> Should we test the interface state/config when there's no peers?

I will add a few nat-lab tests. It make sense.

File: src/device/wg_controller.rs (hunk 4 above)
Comment (LukasPukenis) [reply src/device/wg_controller.rs:200]: Thanks for explaining. It's worth mentioning that on a comment in case other people have the same question in the future :bow:

File: crates/telio-wg/src/adapter/windows_native_wg.rs (hunk 6 above)
Comment (LukasPukenis) [reply crates/telio-wg/src/adapter/windows_native_wg.rs]: > One must set the interface up before configuring peers. And in a mirroring fashion - it is considered "clean" to remove all of them before bringing interface down (in other words, any peer lifetime should be fully covered with interface being up). This is why Down->Up transition is handled before executing Set, and Up->Down after executing Set command

Could it become a comment so it wouldn't be accidentally _refactored_ into what I suggested? :D

Comment (Jauler) [reply crates/telio-wg/src/adapter/windows_native_wg.rs]: TBH this comes to a point where I think it is too cluttery to comment such thing. Eg. its like commenting every `.start()` function in our crates saying that you must first start the module before you can configure it...

File: src/device/wg_controller.rs [hunk 8]
```diff
@@ -183,6 +189,45 @@ async fn consolidate_wg_private_key<W: WireGuard>(
...
+    // down->up transition, listen-port may change. Therefore this consolidation has been added to
+    // keep it in sync all the time.
+
+    if let (None, None) = (proxy, starcast_vpeer) {
+        return Ok(());
+    }
+
+    // Determine WireGuard port. Prefer listen_proxy_port if it is set, otherwise - use
+    // listen_port.
+    // This is only required due to pecuiarities of WireGuard-go bind. Other WireGuard
```
Comment (mathiaspeters) [inline src/device/wg_controller.rs:208]: Nit: peculiarities

Comment (mathiaspeters) [reply src/device/wg_controller.rs:208]: And a question: when we eventually get rid of wireguard-go as a supported adapter type, can we simplify this to just use `listen_port`?

File: nat-lab/tests/test_adapter.py [hunk 9]
```diff
@@ -139,3 +169,121 @@ async def test_adapter_service_loading(
 
     async with AsyncExitStack() as exit_stack:
         _ = await setup_mesh_nodes(exit_stack, [alpha_setup_params, beta_setup_params])
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (Jauler) [reply nat-lab/tests/test_adapter.py:174]: I was thinking, that it would be nice to add at least one test which would go though two cycles, `down->up->down->up`, but did not want to add empty cycling (e.g. on the same scenario or so), this is why I have combined vpn and dns. Meshnet requires slightly different initial setup -> that is why it is separated.

File: src/device/wg_controller.rs (hunk 8 above)
Comment (Jauler) [reply src/device/wg_controller.rs]: yes. We can then eliminate `proxy_listen_port`.

File: nat-lab/tests/test_adapter.py [hunk 10]
```diff
@@ -138,3 +169,121 @@ async def test_adapter_service_loading(
...
+        first_node_id = next(iter(api.nodes))
+        await client_alpha.set_meshnet_config(
+            api.get_meshnet_config(first_node_id, derp_servers=[config.DERP_PRIMARY])
+        )
+
+        # Verify that interface is connected after we have enabled meshnet with peers
+        state = await get_interface_state(client_conn, client_alpha)
+        assert state == AdapterState.UP
+
+        await client_alpha.set_mesh_off()
```
[the remark this answers is not in the document]
Comment (lcruz99) [reply nat-lab/tests/test_adapter.py:285]: If I understood correctly, if the meshnet gets empty and libtelio goes idle, the interface should go down.

[the remark this answers is not in the document]
Comment (Jauler) [reply nat-lab/tests/test_adapter.py:285]: Generally it is equivalent

**Review by lcruz99 — APPROVED** [review]
+1
